File: Kokoro.Graphics.Vulkan/GraphicsPipeline.cpp
#include "GraphicsPipeline.h"
#include "SpecializedShaderModule.h"

Kokoro::Graphics::GraphicsPipeline::GraphicsPipeline() {
	shaders = gcnew List<SpecializedShaderModule^>();
	Topology = TopologyType::Triangle;
	RasterizerDiscard = false;
	LineWidth = 1;
	Cull = CullMode::None;

	ColorBlend = gcnew BlendEqn();
	ColorBlend->DestFactor = BlendFactor::OneMinusSourceAlpha;
	ColorBlend->SrcFactor = BlendFactor::SourceAlpha;
	ColorBlend->Op = BlendOp::Add;

	AlphaBlend = gcnew BlendEqn();
	AlphaBlend->DestFactor = BlendFactor::OneMinusSourceAlpha;
	AlphaBlend->SrcFactor = BlendFactor::SourceAlpha;
	AlphaBlend->Op = BlendOp::Add;

	Fill = FillMode::Fill;

	locked = false;
}

void Kokoro::Graphics::GraphicsPipeline::SetShader(SpecializedShaderModule^ s) {
	shaders->Add(s);
}

void Kokoro::Graphics::GraphicsPipeline::Build(uint32_t w, uint32_t h) {
	if (!locked) {
		VkPipelineVertexInputStateCreateInfo vInputCreatInfo = {};
		vInputCreatInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_VERTEX_INPUT_STATE_CREATE_INFO;
		vInputCreatInfo.vertexBindingDescriptionCount = 0;
		vInputCreatInfo.pVertexBindingDescriptions = nullptr;
		vInputCreatInfo.vertexAttributeDescriptionCount = 0;
		vInputCreatInfo.pVertexAttributeDescriptions = nullptr;

		VkPipelineInputAssemblyStateCreateInfo iaCreatInfo = {};
		iaCreatInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_INPUT_ASSEMBLY_STATE_CREATE_INFO;
		iaCreatInfo.flags = 0;
		iaCreatInfo.topology = TopologyTypeConv::Convert(Topology);
		iaCreatInfo.primitiveRestartEnable = VK_FALSE;

		VkViewport vport = {};
		vport.x = 0.0f;
		vport.y = 0.0f;
		vport.width = static_cast<float>(w);
		vport.height = static_cast<float>(h);
		vport.minDepth = 0.0f;
		vport.maxDepth = 1.0f;

		VkRect2D scissor = {};
		scissor.offset = { 0,0 };
		scissor.extent = {
			w, h
		};

		VkPipelineViewportStateCreateInfo vportCreatInfo = {};
		vportCreatInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_VIEWPORT_STATE_CREATE_INFO;
		vportCreatInfo.viewportCount = 1;
		vportCreatInfo.pViewports = &vport;
		vportCreatInfo.scissorCount = 1;
		vportCreatInfo.pScissors = &scissor;

		VkPipelineRasterizationStateCreateInfo rasCreatInfo = {};
		rasCreatInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_RASTERIZATION_STATE_CREATE_INFO;
		rasCreatInfo.depthClampEnable = VK_FALSE;
		rasCreatInfo.rasterizerDiscardEnable = RasterizerDiscard ? VK_TRUE : VK_FALSE;
		rasCreatInfo.polygonMode = FillModeConv::Convert(Fill);
		rasCreatInfo.lineWidth = LineWidth;
		rasCreatInfo.cullMode = CullModeConv::Convert(Cull);
		rasCreatInfo.frontFace = VK_FRONT_FACE_COUNTER_CLOCKWISE;
		rasCreatInfo.depthBiasEnable = VK_FALSE;
		rasCreatInfo.depthBiasConstantFactor = 0.0f;
		rasCreatInfo.depthBiasClamp = 0.0f;
		rasCreatInfo.depthBiasSlopeFactor = 0.0f;

		VkPipelineMultisampleStateCreateInfo msCreatInfo = {};
		msCreatInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_MULTISAMPLE_STATE_CREATE_INFO;
		msCreatInfo.sampleShadingEnable = VK_FALSE;
		msCreatInfo.rasterizationSamples = VK_SAMPLE_COUNT_1_BIT;
		msCreatInfo.minSampleShading = 1.0f;
		msCreatInfo.pSampleMask = nullptr;
		msCreatInfo.alphaToCoverageEnable = VK_FALSE;
		msCreatInfo.alphaToOneEnable = VK_FALSE;

		VkPipelineDepthStencilStateCreateInfo dpthCreatInfo = {};
		dpthCreatInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_DEPTH_STENCIL_STATE_CREATE_INFO;

		VkPipelineColorBlendAttachmentState colCreatInfo = {};
		colCreatInfo.alphaBlendOp = BlendOpConv::Convert(AlphaBlend->Op);
		colCreatInfo.srcAlphaBlendFactor = BlendFactorConv::Convert(AlphaBlend->SrcFactor);
		colCreatInfo.dstAlphaBlendFactor = BlendFactorConv::Convert(AlphaBlend->DestFactor);
		colCreatInfo.colorBlendOp = BlendOpConv::Convert(ColorBlend->Op);
		colCreatInfo.srcColorBlendFactor = BlendFactorConv::Convert(ColorBlend->SrcFactor);
		colCreatInfo.dstColorBlendFactor = BlendFactorConv::Convert(ColorBlend->DestFactor);
		colCreatInfo.colorWriteMask = VK_COLOR_COMPONENT_R_BIT | VK_COLOR_COMPONENT_G_BIT | VK_COLOR_COMPONENT_B_BIT | VK_COLOR_COMPONENT_A_BIT;
		colCreatInfo.blendEnable = VK_TRUE;

		VkPipelineColorBlendStateCreateInfo colBlendCreatInfo = {};
		colBlendCreatInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_COLOR_BLEND_STATE_CREATE_INFO;

		VkDynamicState dynState = VK_DYNAMIC_STATE_VIEWPORT;
		VkPipelineDynamicStateCreateInfo dynCreatInfo = {};
		dynCreatInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_DYNAMIC_STATE_CREATE_INFO;
		dynCreatInfo.dynamicStateCount = 1;
		dynCreatInfo.pDynamicStates = &dynState;

		VkPipelineLayoutCreateInfo pipelineCreatInfo = {};
		pipelineCreatInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_LAYOUT_CREATE_INFO;
		//pipelineCreatInfo.s
	}
}

File: Kokoro.Graphics/GraphicsPipeline.cs
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using VulkanSharp.Raw;
using static VulkanSharp.Raw.Vk;

namespace Kokoro.Graphics
{
    public enum FillMode
    {
        Fill = VkPolygonMode.PolygonModeFill,
        Line = VkPolygonMode.PolygonModeLine,
        Point = VkPolygonMode.PolygonModePoint
    }

    public class GraphicsPipeline
    {
        public string Name { get; set; }
        public ShaderSource[] Shaders { get; set; }
        public PrimitiveType Topology { get; set; } = PrimitiveType.Triangle;
        public bool DepthClamp { get; set; }
        public bool RasterizerDiscard { get; set; } = false;
        public float LineWidth { get; set; } = 1.0f;
        public CullMode CullMode { get; set; } = CullMode.None;
        public bool EnableBlending { get; set; }
        public DepthTest DepthTest { get; set; } = DepthTest.Greater;
        public FillMode Fill { get; set; } = FillMode.Fill;
        public bool DepthWrite { get; set; } = false;
        public RenderPass RenderPass { get; set; }
        public PipelineLayout PipelineLayout { get; set; }
        public Memory<int>[] SpecializationData { get; set; }
        public bool ViewportDynamic { get; set; }
        public uint ViewportX { get; set; }
        public uint ViewportY { get; set; }
        public uint ViewportWidth { get; set; }
        public uint ViewportHeight { get; set; }
        public float ViewportMinDepth { get; set; }
        public float ViewportMaxDepth { get; set; }

        internal IntPtr hndl;
        private int devID;
        private bool locked;

        public GraphicsPipeline()
        {
        }

        public void Build(int deviceIndex)
        {
            if (!locked)
            {
                unsafe
                {
                    //create pipeline shader stages
                    var shaderStages = new VkPipelineShaderStageCreateInfo[Shaders.Length];
                    var shaderSpecializations = new ManagedPtr<VkSpecializationInfo>[Shaders.Length];
                    var shaderSpecializationBufferPtrs = new MemoryHandle[Shaders.Length];
                    for (int i = 0; i < shaderStages.Length; i++)
                    {
                        if (SpecializationData != null)
                        {
                            shaderSpecializationBufferPtrs[i] = SpecializationData[i].Pin();

                            shaderSpecializations[i] = new VkSpecializationInfo()
                            {
                                mapEntryCount = (uint)SpecializationData[i].Length,
                                pMapEntries = Shaders[i].Specialize(),
                                dataSize = (uint)SpecializationData[i].Length * sizeof(int),
                                pData = (IntPtr)shaderSpecializationBufferPtrs[i].Pointer
                            }.Pointer();
                        }

                        shaderStages[i].sType = VkStructureType.StructureTypePipelineShaderStageCreateInfo;
                        shaderStages[i].stage = (VkShaderStageFlags)Shaders[i].ShaderType;
                        shaderStages[i].module = Shaders[i].ids[deviceIndex];
                        shaderStages[i].pName = "main";
                        shaderStages[i].pSpecializationInfo = shaderSpecializations[i] != null ? shaderSpecializations[i] : IntPtr.Zero;
                    }
                    var shaderStages_ptr = shaderStages.Pointer();

                    //VkPipelineVertexInputStateCreateInfo - dummy, engine doesn't support fix function vertex input
                    var vertexInputInfo = new VkPipelineVertexInputStateCreateInfo()
                    {
                        sType = VkStructureType.StructureTypePipelineVertexInputStateCreateInfo,
                        vertexBindingDescriptionCount = 0,
                        vertexAttributeDescriptionCount = 0,
                    };
                    var vertexInputInfo_ptr = vertexInputInfo.Pointer();

                    //VkPipelineInputAssemblyStateCreateInfo - state
                    var inputAssembly = new VkPipelineInputAssemblyStateCreateInfo()
                    {
                        sType = VkStructureType.StructureTypePipelineInputAssemblyStateCreateInfo,
                        topology = (VkPrimitiveTopology)Topology,
                        primitiveRestartEnable = false
                    };
                    var inputAssembly_ptr = inputAssembly.Pointer();

                    //VkPipelineViewportStateCreateInfo - dynamic state, no scissor
                    var viewport = new VkViewport()
                    {
                        x = ViewportX,
                        y = ViewportY,
                        width = ViewportWidth,
                        height = ViewportHeight,
                        minDepth = ViewportMinDepth,
                        maxDepth = ViewportMaxDepth,
                    };

                    var scissor = new VkRect2D()
                    {
                        offset = new VkOffset2D()
                        {
                            x = (int)ViewportX,
                            y = (int)ViewportY
                        },
                        extent = new VkExtent2D()
                        {
                            width = ViewportWidth,
                            height = ViewportHeight
                        }
                    };

                    var viewport_ptr = viewport.Pointer();
                    var scissor_ptr = scissor.Pointer();
                    var viewportCreateInfo = new VkPipelineViewportStateCreateInfo()
                    {
                        sType = VkStructureType.StructureTypePipelineViewportStateCreateInfo,
                        viewportCount = 1,
                        pViewports = viewport_ptr,
                        scissorCount = 1,
                        pScissors = scissor_ptr
                    };
                    var viewportCreateInfo_ptr = viewportCreateInfo.Pointer();

                    //VkPipelineRasterizationStateCreateInfo - state
                    var rasterizer = new VkPipelineRasterizationStateCreateInfo()
                    {
                        sType = VkStructureType.StructureTypePipelineRasterizationStateCreateInfo,
                        depthClampEnable = DepthClamp,
                        rasterizerDiscardEnable = RasterizerDiscard,
                        polygonMode = (VkPolygonMode)Fill,
                        lineWidth = LineWidth,
                        cullMode = (VkCullModeFlags)CullMode,
                        frontFace = VkFrontFace.FrontFaceCounterClockwise,   //OpenGL default
                        depthBiasEnable = false,
                        depthBiasConstantFactor = 0,
                        depthBiasClamp = 0,
                        depthBiasSlopeFactor = 0,
                    };
                    var rasterizer_ptr = rasterizer.Pointer();

                    //VkPipelineMultisampleStateCreateInfo - don't support for now
                    var multisampling = new VkPipelineMultisampleStateCreateInfo()
                    {
                        sType = VkStructureType.StructureTypePipelineMultisampleStateCreateInfo,
                        sampleShadingEnable = false,
                        rasterizationSamples = VkSampleCountFlags.SampleCount1Bit,
                        minSampleShading = 1,
                        pSampleMask = null,
                        alphaToCoverageEnable = false,
                        alphaToOneEnable = false
                    };
                    var multisampling_ptr = multisampling.Pointer();

                    //VkPipelineDepthStencilStateCreateInfo - state
                    var depthStencil = new VkPipelineDepthStencilStateCreateInfo()
                    {
                        sType = VkStructureType.StructureTypePipelineDepthStencilStateCreateInfo,
                        depthTestEnable = true,
                        depthCompareOp = (VkCompareOp)DepthTest,
                        depthWriteEnable = DepthWrite,
                        depthBoundsTestEnable = false,
                        stencilTestEnable = false,
                        maxDepthBounds = ViewportMaxDepth,
                        minDepthBounds = ViewportMinDepth,
                    };
                    var depthStencil_ptr = depthStencil.Pointer();


                    //VkPipelineColorBlendStateCreateInfo - state
                    var colorBlendStates = new VkPipelineColorBlendAttachmentState[RenderPass.ColorAttachments == null ? 0 : RenderPass.ColorAttachments.Length];
                    for (int i = 0; i < colorBlendStates.Length; i++)
                    {
                        colorBlendStates[i] = new VkPipelineColorBlendAttachmentState()
                        {
                            colorWriteMask = VkColorComponentFlags.ColorComponentRBit | VkColorComponentFlags.ColorComponentGBit | VkColorComponentFlags.ColorComponentBBit | VkColorComponentFlags.ColorComponentABit,
                            blendEnable = false,
                        };
                    }
                    var colorBlendStates_ptr = colorBlendStates.Pointer();

                    var colorBlend = new VkPipelineColorBlendStateCreateInfo()
                    {
                        sType = VkStructureType.StructureTypePipelineColorBlendStateCreateInfo,
                        logicOpEnable = false,
                        logicOp = VkLogicOp.LogicOpCopy,
                        attachmentCount = (uint)colorBlendStates.Length,
                        pAttachments = colorBlendStates_ptr,
                    };
                    colorBlend.blendConstants[0] = 0;
                    colorBlend.blendConstants[1] = 0;
                    colorBlend.blendConstants[2] = 0;
                    colorBlend.blendConstants[3] = 0;
                    var colorBlend_ptr = colorBlend.Pointer();

                    var dynamicStates = stackalloc VkDynamicState[]
                    {
                        VkDynamicState.DynamicStateViewport
                    };

                    var dynamicState = new VkPipelineDynamicStateCreateInfo()
                    {
                        sType = VkStructureType.StructureTypePipelineDynamicStateCreateInfo,
                        dynamicStateCount = ViewportDynamic ? 1u : 0u,
                        pDynamicStates = dynamicStates
                    };
                    var dynamicState_ptr = dynamicState.Pointer();

                    //Setup graphics pipeline
                    var pipelineInfo = new VkGraphicsPipelineCreateInfo()
                    {
                        sType = VkStructureType.StructureTypeGraphicsPipelineCreateInfo,
                        stageCount = (uint)shaderStages.Length,
                        pStages = shaderStages_ptr,
                        pVertexInputState = vertexInputInfo_ptr,
                        pInputAssemblyState = inputAssembly_ptr,
                        pViewportState = viewportCreateInfo_ptr,
                        pRasterizationState = rasterizer_ptr,
                        pMultisampleState = multisampling_ptr,
                        pDepthStencilState = depthStencil_ptr,
                        pColorBlendState = colorBlend_ptr,
                        pDynamicState = dynamicState_ptr,
                        layout = PipelineLayout.hndl,
                        renderPass = RenderPass.hndl,
                        subpass = 0,
                        basePipelineHandle = IntPtr.Zero,
                        basePipelineIndex = -1,
                    };

                    IntPtr pipeline_l = IntPtr.Zero;
                    if (vkCreateGraphicsPipelines(GraphicsDevice.GetDeviceInfo(deviceIndex).Device, IntPtr.Zero, 1, pipelineInfo.Pointer(), null, &pipeline_l) != VkResult.Success)
                        throw new Exception("Failed to create graphics pipeline.");
                    hndl = pipeline_l;

                    if (GraphicsDevice.EnableValidation)
                    {
                        var objName = new VkDebugUtilsObjectNameInfoEXT()
                        {
                            sType = VkStructureType.StructureTypeDebugUtilsObjectNameInfoExt,
                            pObjectName = Name,
                            objectType = VkObjectType.ObjectTypePipeline,
                            objectHandle = (ulong)hndl
                        };
                        GraphicsDevice.SetDebugUtilsObjectNameEXT(GraphicsDevice.GetDeviceInfo(deviceIndex).Device, objName.Pointer());
                    }

                    devID = deviceIndex;
                    locked = true;
                }
            }
            else
                throw new Exception("Pipeline is locked.");
        }
    }
}
